### DISCUSSION TITLE
Counldn't get all tables with base.classes

### DISCUSSION THREAD

**Discussion by cx2810** [sqlalchemy/sqlalchemy#10104] [Usage Questions] [answered]
i use the same code to get all tables in my db, but the results is diff
_the code is run in flask shell_
 ```
from sqlalchemy.ext.automap import automap_base
from sqlalchemy.orm import Session
engine = db.engines.get("my_db")
session = Session(bind=engine)
base = automap_base()
base.prepare(engine, reflect=True)
dir(base.classes)
```
```
python 3.8.8
flask 2.2.2
Flask-SQLAlchemy 3.0.2
SQLAlchemy 1.4.31
```

most of the result is all tables, but only one just contains some tables
i don't know why, please give me some suggestion.
expect your answer.

**Comment by CaselIT** [comment]
Hi,

not sure what's going on. not much information is give.
Try to verify that you are connecting to the same database with the same user

**Reply by cx2810** [reply]
thanks for your reply！
i use the same iso to produce many environments on arm arch, so the db and user are same. but i don't know what happened on one of the environment that the table is not all.
i tried to use the same py script to start a new service, unfortunately i got the same result.
 the original service is running for keeping the wrong env.


what information would be useful？
is there a cache somewhere？

**Reply by CaselIT** [reply]
Can you see the tables if you log in directly to the db?

**Reply by cx2810** [reply]
yes， i can see them

**Reply by CaselIT** [reply]
If you verified that the tables exist in the db and are visible to the user I'm not sure why they would not be loaded.

Try setting the echo parameters of the engine to debug and look at that's reflected to see if all the tables are there

**Reply by cx2810** [reply]
i get all the tables from base.metadata.tables. What's the difference between base.metadata.tables and base.classes

i found this code in sqlalchemy.ext.automap.py in funtion prepare
```
            for table in cls.metadata.tables.values():
                lcl_m2m, rem_m2m, m2m_const = _is_many_to_many(cls, table)
                if lcl_m2m is not None:
                    many_to_many.append((lcl_m2m, rem_m2m, m2m_const, table))
                elif not table.primary_key:
                    continue
                elif table not in table_to_map_config:
                    mapped_cls = type(
                        classname_for_table(cls, table.name, table),
                        (cls,),
                        {"__table__": table},
                    )
                    map_config = _DeferredMapperConfig.config_for_cls(
                        mapped_cls
                    )
                    cls.classes[map_config.cls.__name__] = mapped_cls
                    table_to_map_config[table] = map_config
```
so the problem is why table.primary_key is None

**Reply by CaselIT** [reply, marked as answer]
An orm model requires a primary key, since sqlalchemy needs a way of identifying each row, so ensure that all the tables have a primary key
